### ISSUE TITLE
Bug in checking BMI output for forcing provider interface

### ISSUE THREAD

**Issue by hellkite500** [NOAA-OWP/ngen#393]
# TL;DR
BMI_Module_Formulation forcing provider can self reference a requested output variable to a module's own input variable.

# Background
#387 exposed a subtle bug in the implementation of the forcing provider interface function `get_value`.  The logic previously looked like this:

```c++
double get_value(const std::string &output_name, const time_t &init_time, const long &duration_s,
                         const std::string &output_units) override
        {
            // First make sure this is an available output
            const std::vector<std::string> forcing_outputs = get_available_forcing_outputs();
            if (std::find(forcing_outputs.begin(), forcing_outputs.end(), output_name) == forcing_outputs.end()) {
                throw runtime_error(get_formulation_type() + " received invalid output forcing name " + output_name);
            }
            // TODO: do this, or something better, later; right now, just assume anything using this as a provider is
            //  consistent with times
            /*
            if (last_model_response_delta == 0 && last_model_response_start_time == 0) {
                throw runtime_error(get_formulation_type() + " does not properly set output time validity ranges "
                                                             "needed to provide outputs as forcings");
            }
            */

            // Handle ET requests slightly differently
            //TODO: This should really only happen if neither of these output_names are provided by the module...
            // But this code should also probably be removed in the near future (see GH #297 second comment).
            if (output_name == NGEN_STD_NAME_POTENTIAL_ET_FOR_TIME_STEP || output_name == CSDMS_STD_NAME_POTENTIAL_ET) {
                return calc_et();
            }

            // If not ET, now get correct BMI variable name, which may be the output or something mapped to this output.
            // TODO: move this to a dedicated function
            std::string bmi_var_name;
            std::vector<std::string> output_names = get_bmi_model()->GetOutputVarNames();
            if (std::find(output_names.begin(), output_names.end(), output_name) != output_names.end()) {
                bmi_var_name = output_name;
            }
            else {
                for (auto & iter : bmi_var_names_map) {
                    if (iter.second == output_name) {
                        bmi_var_name = iter.first;
                        break;
                    }
                }
            }
            
            double value = get_var_value_as_double(bmi_var_name);

            // Convert units
            std::string native_units = get_bmi_model()->GetVarUnits(bmi_var_name);
            try {
                return UnitsHelper::get_converted_value(native_units, value, output_units);
            }
            catch (const std::runtime_error& e){
                std::cerr<<"Unit conversion error: "<<std::endl<<e.what()<<std::endl<<"Returning unconverted value!"<<std::endl;
                return value;
            }
        }
```

The problematic logic is the following:
```c++
            // If not ET, now get correct BMI variable name, which may be the output or something mapped to this output.
            // TODO: move this to a dedicated function
            std::string bmi_var_name;
            std::vector<std::string> output_names = get_bmi_model()->GetOutputVarNames();
            if (std::find(output_names.begin(), output_names.end(), output_name) != output_names.end()) {
                bmi_var_name = output_name;
            }
            else {
                for (auto & iter : bmi_var_names_map) {
                    if (iter.second == output_name) {
                        bmi_var_name = iter.first;
                        break;
                    }
                }
            }
```

In #387 this code was refactored into `get_bmi_var_name(...)` and the check for internal ET was done after checking if it was available from a BMI module.

# The bug in action
What is problematic with this is the following scenario:

A realization configuration specifies a BMI model which has the following input variable, such as in CFE
```C
static const char *input_var_names[INPUT_VAR_NAME_COUNT] = {
        "water_potential_evaporation_flux"
};
```
Then this BMI module is configured with the following map
```json
"variables_names_map" : {
       "water_potential_evaporation_flux" : "potential_evapotranspiration"
 }
```

Now consider the following operations
- BMI Module formulation is constructed, and `inner_create_formulation` adds a self reference data provider for these two macros
   - ```c++
      // Supported Standard Names for BMI variables
      // This is needed to provide a calculated potential ET value back to a BMI model
      #define NGEN_STD_NAME_POTENTIAL_ET_FOR_TIME_STEP "potential_evapotranspiration"

      // Taken from the CSDMS Standard Names list
      // TODO: need to add these in for anything BMI model input or output variables we need to know how to recognize
      #define CSDMS_STD_NAME_POTENTIAL_ET "water_potential_evaporation_flux"
      ```
- CFE requests an input of `water_potential_evaporation_flux`
- in `set_model_inputs_prior_to_update(...)` this variable is recognized as an input to the model
- the name map is checked for this variable, and we look through forcing providers for `potential_evapotranspiration` and find the provider (self)
- call `get_value("potential_evapotranspiration")`

Now prior to #387  this would always hit the `calc_et()` conditional and early return from the function.  But other variables could potentially have triggered this bug.  Now with the refactored code, we have the following sequence of events.

- Validate `potential_evapotranspiration` is an available forcing output (it is...)
- Check if `potential_evapotranspiration` is a BMI variable
   - look in `get_bmi_model()->GetOutputVarNames()`, don't find it
   - look in `bmi_var_names_map`
      - Find the mapped name (iter.second == "potential_evapotranspiration)
      - set `bmi_var_name = iter.first`, which in this case is `water_potential_evaporation_flux`
- So now we have a `bmi_var_name = water_potential_evaporation_flux`
- Proceed to `get_var_value_as_double(bmi_var_name)`

We have effectively self referenced a requested output variable to an input variable!

# Solution
Make sure `get_bmi_var_name` validates mapped entries as output variables, since for the forcing provider interface, this is the intended semantic.

```c++
        inline void get_bmi_var_name(const std::string &name, std::string &bmi_var_name)
        {
            //check standard output names first
            std::vector<std::string> output_names = get_bmi_model()->GetOutputVarNames();
            if (std::find(output_names.begin(), output_names.end(), name) != output_names.end()) {
                bmi_var_name = name;
            }
            else
            {
                //check mapped names
                std::string mapped_name;
                for (auto & iter : bmi_var_names_map) {
                    if (iter.second == name) {
                        mapped_name = iter.first;
                        break;
                    }
                }
                //ensure mapped name maps to an output variable
                if (std::find(output_names.begin(), output_names.end(), mapped_name) != output_names.end()){
                    bmi_var_name = mapped_name;
                }
                //else not an output variable
            }
        }
```

**Comment by hellkite500** [comment]
It also occurs to me that it might be good to rename `get_bmi_var_name` to `get_bmi_output_var_name` to be quite clear what is supposed to happen using this logic.
